File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */

// null, 3, null, 2, 1
// null, 1, 3, 2, null
// 1, null,2,3,null/
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        this.inorder(root, res);
        return res;
    }

    public void inorder(TreeNode t, List<Integer> arr) {
        if (t == null)
            return;

        inorder(t.left, arr);
        arr.add(t.val);
        inorder(t.right, arr);
    }
}

File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */

//  LRootR
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        TreeNode* temp = root;
        while (temp != NULL) {
            if (temp->left==NULL){
                res.push_back(temp->val);
                temp=temp->right;
            }else {
                TreeNode* t2=temp->left;
                while(t2->right!=NULL&&t2->right!=temp){
                    t2=t2->right;
                }
                if(t2->right==NULL){
                    t2->right=temp;
                    temp=temp->left;
                }
                else{
                    res.push_back(temp->val);
                    t2->right=NULL;
                    temp=temp->right;
                }
            }
        }
        return res;
    }
};